### ISSUE TITLE
mat: revisit Errors and panic handling

### ISSUE THREAD

**Issue by kortschak** [gonum/gonum#993]
In the initial design of the matrix package I included the error type and maybe function. There was an intention that this could be used by client code to recover from errors in a way that mimics error handling without the possibility of clients ignoring returned errors.

In that design there are two types of panics:

1. panics that are client errors
2. panics that are our errors

When we panic internally, we use strings, rather than the `Error` type so that the panic does propagate to cause the application to panic completely.

However, it looks like this has not been consistently used. We do consistently use the named errors, but I think there are errors that are user errors that we panic with strings.

This issue is to decide whether to keep the distinction between the two types of error that we panic with and fix up the discordant string panics, or make all panics the same type (or some third option).

**Comment by kortschak** [comment]
As of 39b295f7af12a6144c9f98b1d0a5cb797d638468 these are the naked string panics.

I think the "negative dimension" use and "caps not correctly set" are fine. The two "unreachable" cases should probably be "invalid norm". The Eigen size mismatch is internal and obviously correct. I'll go through more later with detail.

- [ ] band.go:134: panic("mat: negative dimension")
- [ ] band.go:137: panic("mat: band out of range")
- [ ] cdense.go:41: panic("mat: negative dimension")
- [ ] cdense.go:68: panic("mat: caps not correctly set")
- [ ] cdense.go:93: panic("mat: caps not correctly set")
- [ ] cholesky.go:163: panic("cholesky: matrix must be upper triangular")
- [ ] cholesky.go:420: panic("cholesky: scaling by a non-positive constant")
- [ ] dense_arithmetic.go:629: panic("matrix: illegal power")
- [ ] dense_arithmetic.go:826: panic("mat: caps not correctly set")
- [ ] dense.go:50: panic("mat: negative dimension")
- [ ] dense.go:77: panic("mat: caps not correctly set")
- [ ] dense.go:106: panic("mat: caps not correctly set")
- [ ] diagonal.go:84: panic("mat: negative dimension")
- [ ] eigen.go:286: panic("size mismatch")
- [ ] eigen.go:289: panic("size mismatch")
- [ ] errors.go:55: panic("mat: invalid error")
- [ ] errors.go:77: panic("mat: invalid error")
- [ ] errors.go:98: panic("mat: invalid error")
- [ ] gsvd.go:94: panic("gsvd: bad input kind")
- [ ] gsvd.go:340: panic("mat: improper GSVD kind")
- [ ] gsvd.go:369: panic("mat: improper GSVD kind")
- [ ] gsvd.go:398: panic("mat: improper GSVD kind")
- [ ] hogsvd.go:54: panic("hogsvd: too few matrices")
- [ ] hogsvd.go:169: panic("hogsvd: invalid index")
- [ ] hogsvd.go:199: panic("hogsvd: invalid index")
- [ ] io.go:144: panic("mat: unmarshal into non-zero matrix")
- [ ] io.go:203: panic("mat: unmarshal into non-zero matrix")
- [ ] io.go:330: panic("mat: unmarshal into non-zero vector")
- [ ] io.go:383: panic("mat: unmarshal into non-zero vector")
- [ ] matrix.go:364: panic("mat: bad norm value")
- [ ] matrix.go:771: panic("unreachable")
- [ ] matrix.go:790: panic("unreachable")
- [ ] svd.go:195: panic("svd: u not computed during factorization")
- [ ] svd.go:229: panic("svd: v not computed during factorization")
- [ ] symband.go:86: panic("mat: negative dimension")
- [ ] symband.go:89: panic("mat: band out of range")
- [ ] symband.go:155: panic("mat: blas64.SymmetricBand does not have blas.Upper storage")
- [ ] symmetric.go:68: panic("mat: negative dimension")
- [ ] triangular.go:126: panic("mat: negative dimension")
- [ ] triangular.go:196: panic("mat: cannot convert unit TriDense into blas64.Symmetric")
- [ ] triangular.go:227: panic("mat: cannot set TriDense with Unit storage format")
- [ ] triangular.go:328: panic("mat: cannot take view of Unit diagonal")
- [ ] triangular.go:568: panic("mat: triangle size mismatch")
- [ ] triband.go:202: panic("mat: negative dimension")
- [ ] triband.go:205: panic("mat: band out of range")
- [ ] triband.go:331: panic("mat: cannot set TriBand with Unit storage")
- [ ] triband.go:339: panic("mat: cannot take view of Unit diagonal")
- [ ] vector.go:94: panic("mat: negative dimension")

**Comment by kortschak** [comment]
ISTM that the following should be `mat.Error`:

- "mat: band out of range"
- "cholesky: matrix must be upper triangular"
- "cholesky: scaling by a non-positive constant"
- "matrix: illegal power" (should also be "mat:...")
- "gsvd: bad input kind"/"mat: improper GSVD kind" (same text?)
- "hogsvd: too few matrices"
- "hogsvd: invalid index"
- "mat: bad norm value"
- "svd: u not computed during factorization"/"svd: v not computed during factorization" (also probably `badFact` and `badNoVect` should be `mat.Error` - more below)
- "mat: blas64.SymmetricBand does not have blas.Upper storage"
- "mat: cannot set TriDense with Unit storage format"
- "mat: cannot take view of Unit diagonal"
- "mat: cannot set TriBand with Unit storage"
- "mat: cannot take view of Unit diagonal"

The "mat: unmarshal into non-zero matrix" panic is on error-returning methods. Maybe these should just be an error.

"mat: cannot convert unit TriDense into blas64.Symmetric" and "mat: triangle size mismatch" are only used internally.

I'm not sure about "mat: negative dimension", but it seems of the same order as `ErrShape`, so probably it should be `mat.Error` as well.

As with `badFact`, and `badNoFact`, these should be elevated to `mat.Error`: `badCholesky`, `badSliceLength`, `badLQ`, `badLU`, `badSymTriangle` (not in `reuseAs` since this is an internal use) and `badQR`.

**Comment by friend0** [comment]
I found this issue searching for 'Panic' as I've got some similar concerns about error handling and panics in this library.

I think it's an unspoken go standard to never let a panic leave your code unhandled. In the current matrix implementations, the
documentation enumerated the conditions under which the function will panic. This is inherently less clear than returning an error like a majority of standard go libraries. The understanding is that any function that can panic or throw an error will return that error so that the user can handle it explicitly. 

As it is now, I have to be looking at the documentation of the function to see that I'll need to write my own wrapper to handle panics (or worse, I encounter errors at runtime) vs an error returning function that will light up any IDE with a go LSP if it goes unhandled. 

Is there a world where gonum will not panic on routine errors?

**Comment by kortschak** [comment]
That behaviour will not change. The decision about panic versus error comes down to these things:

1. People don't check errors, particularly when the only returned value is an error as is the case with the matrix operations.
2. Errors in the domain that the libraries were written for *must* be handled (unsafe scientific computing was a motivating reason for starting the Gonum project).
3. Errors in code fall into two classes: those that can be known/prevented ahead of time; and those that cannot.

Given 1. and 2. we need to have a way to enforce error checking. A panic is the only way that we can do this given the type system limitations in the language. When an error condition arises due to incorrect programming, a panic will happen, but when an error is something that could be expected to happen even when correct checks have been performs (say the matrix is singular for example), then a go `error` value will be returned. A good way of thinking about panics in Gonum is that it is an extension of type checking into runtime because compile-time type checking is not possible.

tl;dr; Make the code correct and you won't need to recover panics.

**Comment by friend0** [comment]
Thank you for the response, I can see where you're coming from. I do disagree on the point that people don't check their errors in Go, as this is a baseline expectation. Any modern development environment - and even the more arcane variety - will let you know that you're failing to check errors. A panic is not really the only way to ensure this, as many methods are already returning pointers. A nil pointer and an error are a sure fire way to get you to check the error, and to write the code correctly. In any case, it's the only show in town as far as Go is concerned, so I abide.

**Comment by kortschak** [comment]
> A *nil pointer* and an error are a sure fire way to get you to check the error

This is the issue (my emphasis). Where is the nil pointer in the following code?
```
a.Mul(b, c)
```
I could imagine (and see regularly) cases where error-only returning functions do not have their errors checked. In many situations shoddy code like that is acceptable because it just means that a page doesn't load, or someones data get's dropped on the floor during an e-commerce transaction. The basis for Gonum was to support safe scientific coding with real-world harm impacts, so we cannot trust that someone's good coding practices will be exerted.

The shape of a matrix in  mathematical context is essentially part of its type. This cannot be expressed in the Go type system, but it is conceptually equivalent to `var b bool; var i int; c := b * i`, which would obviously be caught at compile time. If we could catch this kind of thing at compile time, we would, but that would not be Go.
